heuristic/delta_calculations.py

```diff
@@ -56,212 +42,128 @@ def calculate_weekly_rest(state, shifts_at_week, employees, weeks):
-                (important[week], float((important[week + 1] - important[week])))
-            ]
+        if(len(actual_shifts[key]) == 0):
+            off_shift_periods[key] = [(important[week], float((important[week + 1] - important[week])))]
 
         else:
-            if actual_shifts[key][0][0] - important[week] >= 36:
-                off_shift_periods[key].append(
-                    (important[week], actual_shifts[key][0][0] - important[week])
-                )
+            if(actual_shifts[key][0][0] - important[week] >= 36):
+                off_shift_periods[key].append((important[week], actual_shifts[key][0][0] - important[week]))
 
-            if important[week + 1] - (actual_shifts[key][-1][0] + actual_shifts[key][-1][1]) >= 36:
-                off_shift_periods[key].append(
-                    (
-                        (actual_shifts[key][-1][0] + actual_shifts[key][-1][1]),
-                        important[week + 1]
-                        - (actual_shifts[key][-1][0] + actual_shifts[key][-1][1]),
-                    )
-                )
+            if(important[week + 1] - (actual_shifts[key][-1][0] + actual_shifts[key][-1][1]) >= 36):
+                off_shift_periods[key].append(((actual_shifts[key][-1][0] + actual_shifts[key][-1][1]), important[week + 1] - (actual_shifts[key][-1][0] + actual_shifts[key][-1][1])))
+
+            for i in range(len(actual_shifts[key])-1):
+                if(actual_shifts[key][i+1][0] - (actual_shifts[key][i][0] + actual_shifts[key][i][1]) >= 36):
+                    off_shift_periods[key].append(((actual_shifts[key][i][0] + actual_shifts[key][i][1]), actual_shifts[key][i+1][0] - (actual_shifts[key][i][0] + actual_shifts[key][i][1])))
 
-            for i in range(len(actual_shifts[key]) - 1):
-                if (
-                    actual_shifts[key][i + 1][0]
-                    - (actual_shifts[key][i][0] + actual_shifts[key][i][1])
-                    >= 36
-                ):
-                    off_shift_periods[key].append(
-                        (
-                            (actual_shifts[key][i][0] + actual_shifts[key][i][1]),
-                            actual_shifts[key][i + 1][0]
-                            - (actual_shifts[key][i][0] + actual_shifts[key][i][1]),
-                        )
-                    )
-
-        if len(off_shift_periods[key]) != 0:
-            state.w[key] = max(off_shift_periods[key], key=itemgetter(1))
+        if(len(off_shift_periods[key]) != 0):
+            state.w[key] = max(off_shift_periods[key],key=itemgetter(1))
             state.hard_vars["weekly_off_shift_error"][key] = 0
         else:
             state.hard_vars["weekly_off_shift_error"][key] = 1
             state.w[key] = (0, 0.0)
 
+        
 
-# Weaknesses:
+#Weaknesses: 
 # 1. It checks every weekend instead of only the weekends that have been destroyed or repaired
 def calculate_partial_weekends(state, employees, shifts_at_day, saturdays):
```

Comment:
I think we should use `weeks` as an index for partial weekends, and not `saturdays`. By doing so I think it will be much easier to only loop over the affected weekends. In addition, we can merge the looping in some case, e. g. 
```python
sum(w[e, j][1] for j in data["time"]["weeks"])
- sum(soft_vars["deviation_contracted_hours"][e, j] for j in data["time"]["weeks"])
- sum(soft_vars["partial_weekends"][e, i] for i in data["time"]["saturdays"])
```
can be written as: 
```python
sum(
    w[e, j][1] 
    - soft_vars["deviation_contracted_hours"][e, j]
    - soft_vars["partial_weekends"][e, i]
    for j in data["time"]["weeks"]
)
```
and thus reducing the number of necessary loops. 

If you agree, I can look into it over the weekend.

Reply:
I could be a good idea, but it is important to be careful when changing indexes as they are used a lot of places. Have to be absolutely sure that it is possible to do it for every place it is used. Even though this loop might take shorter time another loop might be destroyed or take longer time.